### DISCUSSION TITLE
Openthread LinkQualityInfo->mMessageErrorRate

### DISCUSSION THREAD

**Discussion by farhangj** [openthread/openthread#9135] [Q&A] [answered]
Hi,

I have a network of 1 or multiple routers and bunch of SEDs based on nRF NCS 1.9/Zephyr 2.7.99.

I'm capturing timeseries of the messageErrorRate property and I see it go up and down in large steps.
It goes up from 0% to 99% (once converted to percentage) and back down from 99% to 0%. I have seen values in between at times, but it seems like it is spending most of its time either at 0 or 99%. Very non-linear/twitchy behavior.
the value is converted to percentage using this math:

uint8_t msg_error_p = (uint8_t)(((float)(neighbor_info.mMessageErrorRate*100.0))/(float)(0xffff));

based on this line in thread.h:
uint16_t     mMessageErrorRate;     ///< (IPv6) msg error rate **(0xffff->100%)**. Requires **error tracking feature.**

While it is reported to be 99% for a particular MTD/ED, it is surprising that my custom messages to EDs make it there and replied back over CoAP/UDP just fine. The number of repeats I believe is 3 by default. I'm not seeing nearly as much as 99% message loss.
The poll period of my EDs are at 1s.

- What is the Openthread "error tracking feature"? how do I enable it?
- is mMessageErrorRate an experimental feature?
- How is calculated over time? Does it get calculate as unsuccessful IP packets over entire IP packets count? What interval does it reset at or is it a moving window?
- Could it be that polling messages from SEDs occuring at 1s are causing collisions that cause this high message error rate? Largest setup has 2x routers each with ~8 SEDS polling at 1s.
- Seems like I can correlate this issue to when I increased the TX power of my EDs "to be safer". Is there a feature in OT that can control TX power of EDs automatically based on link quality? So that the developer doesn't have to go with a 1 fits all TX power for OT nodes?

Thanks,

**Comment by abtink** [comment, marked as answer]
@farhangj some answers below:

> What is the Openthread "error tracking feature"? how do I enable it?

It is the feature to track frame and (IPv6) message error rate as part of "link quality" info per neighbor.
It is always enabled 
- We used to have a `CONFIG` feature (`OPENTHREAD_CONFIG_ENABLE_TX_ERROR_RATE_TRACKING` to enable/disable this which was removed in #4417 (we may want to update the `thread.h` API doc to clarify this).

> is mMessageErrorRate an experimental feature?

No. It is been in use for some time. Note that it is a feature of OpenThread (Thread spec does not explicitly define or require this).

> How is calculated over time? Does it get calculate as unsuccessful IP packets over entire IP packets count? What interval does it reset at or is it a moving window?

It uses an exponential moving IIR averaging filter (with a specified weight) and is implemented in [`SuccessRateTracker` class](https://github.com/openthread/openthread/blob/98fe3f3875ce8ecacdc650ff936000ab7eb99e29/src/core/thread/link_quality.hpp#L57-L98). The weights used for different param is defined in `config/link_quality.h`:
https://github.com/openthread/openthread/blob/98fe3f3875ce8ecacdc650ff936000ab7eb99e29/src/core/config/link_quality.h#L38-L75

For converting from `uint16_t` representation you can also consider using something similar to CLI `PercentageToString()` (which avoids floating point calculation):
https://github.com/openthread/openthread/blob/98fe3f3875ce8ecacdc650ff936000ab7eb99e29/src/cli/cli_output.cpp#L265-L274


> Could it be that polling messages from SEDs occuring at 1s are causing collisions that cause this high message error rate? Largest setup has 2x routers each with ~8 SEDS polling at 1s.

Not sure what is causing this on your device. May be worth investigating this.

> Is there a feature in OT that can control TX power of EDs automatically based on link quality? 

No such feature in OT and/or Thread spec today. 

There are proposals under discussion in Thread group for tx power control in next/future versions of Thread spec (1.4 or later). 
My personal view is that this can have its own downsides and requires further analysis and consideration (specially allowing tx power to be reduced can make CSMA less effective and lead to more collisions and frame drops).
